File: amplify/functions/resizeImage/handler.ts
import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
import sharp from 'sharp';

export const handler = async (event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> => {
    try {
        // Parse incoming JSON to extract the base64 image string.
        const { imageBase64 } = JSON.parse(event.body || '{}');
        if (!imageBase64) {
            return { statusCode: 400, body: JSON.stringify({ error: "Missing 'imageBase64' in request body" }) };
        }

        // Convert base64 to Buffer.
        const inputBuffer = Buffer.from(imageBase64, 'base64');

        // Use sharp to get the image metadata.
        const imageSharp = sharp(inputBuffer);
        const metadata = await imageSharp.metadata();
        if (!metadata.width || !metadata.height) {
            throw new Error("Unable to determine image dimensions.");
        }

        console.log(`Original image size: ${metadata.width} x ${metadata.height}`);

        // Determine new dimensions such that the smallest edge becomes 512px.
        let newWidth = metadata.width;
        let newHeight = metadata.height;
        if (metadata.width <= metadata.height && metadata.width !== 512) {
            newWidth = 512;
            newHeight = Math.round((512 * metadata.height) / metadata.width);
        } else if (metadata.height < metadata.width && metadata.height !== 512) {
            newHeight = 512;
            newWidth = Math.round((512 * metadata.width) / metadata.height);
        }

        console.log(`Resized image size: ${newWidth} x ${newHeight}`);

        // Resize the image.
        const resizedBuffer = await imageSharp
            .resize(newWidth, newHeight)
            .png()  // output as PNG
            .toBuffer();

        // Convert the resized image to a base64 string.
        const resizedBase64 = resizedBuffer.toString('base64');

        return {
            statusCode: 200,
            body: JSON.stringify({ imageBase64: resizedBase64 }),
        };
    } catch (error: any) {
        console.error('Resize error:', error);
        return { statusCode: 500, body: JSON.stringify({ error: error.message }) };
    }
};

File: amplify/functions/resizeImage/handler.py
import json
import base64
from io import BytesIO
from PIL import Image

def handler(event, context):
    try:
        # Parse the incoming request body as JSON.
        body = json.loads(event.get('body', '{}'))
        image_base64 = body.get('imageBase64')
        if not image_base64:
            return {
                "statusCode": 400,
                "body": json.dumps({"error": "Missing 'imageBase64' in request body"})
            }
        
        # Decode the base64 string into bytes.
        image_data = base64.b64decode(image_base64)
        with BytesIO(image_data) as input_buffer:
            with Image.open(input_buffer) as img:
                width, height = img.size
                print(f"Original image size: {width}x{height}")

                # Calculate new dimensions so that the smallest edge is 512 pixels.
                if width <= height:
                    new_width = 512
                    new_height = int(height * (512 / width))
                else:
                    new_height = 512
                    new_width = int(width * (512 / height))
                
                print(f"Resized image size: {new_width}x{new_height}")

                # Resize the image using a high-quality filter.
                resized_img = img.resize((new_width, new_height), Image.LANCZOS)

                # Save the resized image to an in-memory buffer in PNG format.
                output_buffer = BytesIO()
                resized_img.save(output_buffer, format='PNG')
                output_data = output_buffer.getvalue()
        
        # Encode the resized image back to a base64 string.
        resized_base64 = base64.b64encode(output_data).decode('utf-8')
        return {
            "statusCode": 200,
            "body": json.dumps({"imageBase64": resized_base64})
        }
    except Exception as e:
        print(f"Error during resize: {str(e)}")
        return {
            "statusCode": 500,
            "body": json.dumps({"error": str(e)})
        }
